src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/srm/matcher/SymbolicRegexMatcher.cs

```diff
@@ -451,14 +412,38 @@ private State<S> Delta(string input, int i, State<S> q)
         {
             int c = input[i];
             // atom_id = atoms.Length represents \Z (last \n)
-            int atom_id = (c == 10 && i == input.Length - 1 && q.Node.info.StartsWithLineAnchor ? atoms.Length : dt.Find(c));
-            int offset = (q.Id << K) | atom_id;
-            var p = delta[offset];
-            if (p == null)
-                //transition atom False means that this is \Z
-                return CreateNewTransition(q, atom_id == atoms.Length ? builder.solver.False : atoms[atom_id], offset);
+            int atom_id = (c == 10 && i == input.Length - 1 && q.StartsWithLineAnchor ? builder.atoms.Length : dt.Find(c));
+            // atom=False represents \Z
+            S atom = atom_id == builder.atoms.Length ? builder.solver.False : builder.atoms[atom_id];
+            if (antimirov && q.Node.Kind == SymbolicRegexKind.Or)
```

Comment:
We might want to check how this extra branch in our innermost matching loop affects the performance of the non-antimirov mode. If it does, then we could bubble up this decision to outside the matching loop. In C++ we could use a bool template parameter to avoid copying code, but I'm not sure if C# has any way to get a similar effect. So we might need to specialize some code by hand.

Reply:
I can set the antimirov flag **in the builder** -- I'll do that -- then it will not be passed around as an extra argument at all in the rest of the code and will only be checked when node.kind is Or (I guess an overhead that should be negligible). I agree with c++ -- templates can nicely solve the issue, but throwing an exception and later using a separate loop for Antimirov mode in C# will cause a lot of code duplication because much of the code will be essentially identical.

Reply:
Ok! I'll rerun one of the microbenchmarks to see how this fares.